### DISCUSSION TITLE
Plot generated using Plot_Data_Plane is reversed

### DISCUSSION THREAD

**Discussion by jwraby** [dtcenter/METplus#2084] [Plot Generation] [answered]
I generated a plot using Python embedding to run Plot_Data_Plane. The Python script which produces the output ndarray and is passed in memory to Plot_Data_Plane has the following grid specifications:
'grid': {
      'type': 'Lambert Conformal',
      'hemisphere': 'N',
      'name': 'lambert',
      'scale_lat_1': 25.0,
      'scale_lat_2': 25.0,
      'lat_pin': 31.277,
      'lon_pin': -107.342,
      'x_pin': 0.0,
      'y_pin': 0.0,
      'lon_orient': -106.0,
      'd_km': 0.500,
      'r_km': 6367.47,
      'nx': 131,
      'ny': 132,

The plots which are produced by the Python script show that the temperature field is reversed from north to south. I checked the accuracy of the Python plots based on an independent set of ground truth data. I 've been testing various modifications of the grid specs, but I can't seem to identify the source of the error which is causing the reversal in the plot produced by Plot_Data_Plane. I will work on uploading examples of the two plots.

**Comment by jwraby** [comment]
The following image was produced by the Python script and accurately shows the plot of 2m temperature:
![Fig1_2maglTMP_anal_5](https://user-images.githubusercontent.com/87020191/225025167-162e8389-4aff-464b-98f4-03ae7f392080.PNG)
The following image was captured from the output .ps file produced by Plot_Data_Plane and shows the N-S reversal:
![2maglTMP_anal_5](https://user-images.githubusercontent.com/87020191/225025909-411c248a-8fa3-4cc1-afbb-e464e1e8b8f7.PNG)

I compared the plots to independent ground truth data and confirmed that the plot produced by the Python script is accurate. The plot produced by Plot_Data_Plane accurately places the data along the eastern and western domain borders, but reverses the data in the north-south sense. For example, the large area of the coldest temperatures is correctly placed on the eastern side of the plot, but it should be in the NE corner instead of the SE corner.

**Comment by jprestop** [comment, marked as answer]
Hi @jwraby. 

I see that you are having trouble with plot_data_plane reversing the data.  

While python embedding provides flexibility, the downside is that it places the onus on the user to ensure that the location and order of the data matches the grid definition. Running plot_data_plane is exactly the right thing to do.

Please take a look at the python embedding examples on the bottom of this page:
https://dtcenter.org/community-code/model-evaluation-tools-met/sample-analysis-scripts

Please download and run the 5th example:

> Plot the PSFC variable from a post-processed WRF output file on a Lambert Conformal projection:
> plot_data_plane PYTHON_NUMPY [psfc.ps](https://dtcenter.org/sites/default/files/community-code/met/python-scripts/psfc.png) 'name="[read_PostProcessed_WRF.py](https://dtcenter.org/sites/default/files/community-code/met/python-scripts/read_PostProcessed_WRF.py.txt) 
> [2004-09-01_04-00-00.nc](https://dtcenter.org/sites/default/files/community-code/met/python-scripts/2004-09-01_04-00-00.nc) PSFC";'
> 

Please note this line from the read_PostProcessed_WRF.py script, which reverses the order of one of the dimensions:
```
met_data = data[::-1].copy()
```

That should help resolve this issue.  Please let us know if you have any other questions.

**Reply by jwraby** [reply]
Thank you for your suggestion and the information needed to try it. This looks very helpful. I will try it soon.

**Reply by jwraby** [reply]
Hi Julie. I was able to run the 5th example test case and successfully produced the output which looks identical to the example image. See attached. I will work on adding the command you pointed out into my Python script and testing it. If I want to run Grid-Stat instead of Plot_Data_Plane would I still need to reverse the order of one of the dimensions by using the same command?
![My_psfc](https://user-images.githubusercontent.com/87020191/225358645-c488f950-4b5f-49d3-883d-4da753b27dfb.PNG)

**Reply by jprestop** [reply]
@jwraby

> If I want to run Grid-Stat instead of Plot_Data_Plane would I still need to reverse the order of one of the dimensions by using the same command?

I believe that is correct.  Let us know if you experience any problems.

**Reply by jwraby** [reply]
I added the command you pointed out to reverse the order of one of the dimensions into my Python script and Plot_Data_Plane successfully produced the expected plot with not reversal. Thanks for your accurate diagnosis of the cause and the resolution of the problem! I will work on a test to see if I can run Grid-Stat with the same output from my Python script using Python embedding.
